### PR TITLE
Refactors how conversations are stored in ConversationsProvider.tsx

### PR DESCRIPTION
**Pull request by rhonsby**
### Description

This is part 2 of refactoring `ConversationsProvider.tsx`. This commit specifically refactors how conversations are stored at the top level into an `inboxes` bucket. This will make it easier for us to continue to adding new inboxes.

### Issue

https://github.com/papercups-io/papercups/issues/760

### Screenshots

https://user-images.githubusercontent.com/1361509/116147217-722c7000-a6ad-11eb-8d54-affa209f3255.mp4

## Checklist

- [ ] Everything passes when running `mix test`
- [ ] Ran `mix format`
- [ ] No frontend compilation warnings

### CODE REVIEW

File: assets/src/components/conversations/ConversationsProvider.tsx [hunk 1]
```diff
@@ -16,20 +16,55 @@ import {
 } from '../../presence';
 import ConversationNotificationManager from './ConversationNotificationManager';
 
+type Inbox = {
+  conversationIds: string[];
+  unreadCount: number;
+};
+
+type Inboxes = {
```
Comment (rhonsby) [inline assets/src/components/conversations/ConversationsProvider.tsx:24]: This structure will make it easier for us to expand to adding custom inboxes as well. 

* `all` represents all of the inboxes we support today.
* `bySource` will contain all of the inboxes that are channel specific.
* We can expand by adding something like `custom` which will contain all of the inboxes that the user configured themselves.

File: assets/src/components/conversations/ConversationsProvider.tsx [hunk 2]
```diff
@@ -374,19 +393,8 @@ export class ConversationsProvider extends React.Component<Props, State> {
...
-        const left = a.last_activity_at
-          ? +new Date(a.last_activity_at)
-          : -Infinity;
-        const right = b.last_activity_at
-          ? +new Date(b.last_activity_at)
-          : -Infinity;
-
-        return right - left;
-      }
+    const sortedConversationIds = this.getSortedConversationIds(
```
Comment (rhonsby) [inline assets/src/components/conversations/ConversationsProvider.tsx:396]: Note that I migrated the logic for sorting conversations into `getSortedConversationIds`.

File: assets/src/components/conversations/ConversationsProvider.tsx [hunk 3]
```diff
@@ -417,47 +421,11 @@ export class ConversationsProvider extends React.Component<Props, State> {
...
-    const mine = openConversations
-      .filter((c) => c.assignee_id === currentUserId && c.status === 'open')
-      .map((c) => c.id);
-    const priority = openConversations
-      .filter((c) => c.priority === 'priority' && c.status === 'open')
-      .map((c) => c.id);
-    const closed = allConversations
-      .filter((c) => c.status !== 'open')
-      .map((c) => c.id);
+    const inboxes = this.getInboxes(updatedConversationsById);
```
Comment (rhonsby) [inline assets/src/components/conversations/ConversationsProvider.tsx:424]: I've noticed that our codebase has a few pretty big functions here and there. I started to clean this up by creating a lot of smaller functions. Let me know what your thoughts are on many small functions vs fewer larger functions.

Comment (reichert621) [reply assets/src/components/conversations/ConversationsProvider.tsx:409]: yup i prefer this as well :+1:

**Review by reichert621 — CHANGES_REQUESTED** [review]
just a couple small things i noticed while testing the branch!

File: assets/src/components/Dashboard.tsx [hunk 4]
```diff
@@ -441,6 +496,15 @@ const Dashboard = (props: RouteComponentProps) => {
             path="/conversations/priority"
             component={PriorityConversations}
           />
+          <Route path="/conversations/live-chat">
+            <ConversationsBySource title="Live chat" source="chat" />
+          </Route>
+          <Route path="/conversations/email">
+            <ConversationsBySource title="Email" source="email" />
```
[the remark this answers is not in the document]
Comment (rhonsby) [reply assets/src/components/Dashboard.tsx:503]: It looks like adding `key` to the `Route` fixed the issue. Can you make sure that it works for you too?

[the remark this answers is not in the document]
Comment (reichert621) [reply assets/src/components/Dashboard.tsx:503]: seems to work for me too! :+1:

**Review by reichert621 — APPROVED** [review]

**Pull request merged**